stl/inc/optional

```diff
@@ -445,6 +469,170 @@ public:
         return static_cast<remove_cv_t<_Ty>>(_STD forward<_Ty2>(_Right));
     }
 
+#if _HAS_CXX23
+    template <class _Fn>
+    constexpr auto and_then(_Fn&& _Func) & {
+        using _Uty = invoke_result_t<_Fn, _Ty&>;
+
+        static_assert(_Is_specialization_v<remove_cvref_t<_Uty>, optional>,
+            "optional<T>::and_then(F) requires the return type of F to be a specialization of optional "
+            "(N4901 [optional.monadic]/2).");
+
+        if (this->_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), this->_Value);
+        } else {
+            return remove_cvref_t<_Uty>{};
+        }
+    }
+
+    template <class _Fn>
+    constexpr auto and_then(_Fn&& _Func) const& {
+        using _Uty = invoke_result_t<_Fn, const _Ty&>;
+
+        static_assert(_Is_specialization_v<remove_cvref_t<_Uty>, optional>,
+            "optional<T>::and_then(F) requires the return type of F to be a specialization of optional "
+            "(N4901 [optional.monadic]/2).");
+
+        if (this->_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), this->_Value);
+        } else {
+            return remove_cvref_t<_Uty>{};
+        }
+    }
+
+    template <class _Fn>
+    constexpr auto and_then(_Fn&& _Func) && {
+        using _Uty = invoke_result_t<_Fn, _Ty>;
+
+        static_assert(_Is_specialization_v<remove_cvref_t<_Uty>, optional>,
+            "optional<T>::and_then(F) requires the return type of F to be a specialization of optional "
+            "(N4901 [optional.monadic]/5).");
+
+        if (this->_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), _STD move(this->_Value));
+        } else {
+            return remove_cvref_t<_Uty>{};
+        }
+    }
+
+    template <class _Fn>
+    constexpr auto and_then(_Fn&& _Func) const&& {
+        using _Uty = invoke_result_t<_Fn, const _Ty>;
+
+        static_assert(_Is_specialization_v<remove_cvref_t<_Uty>, optional>,
+            "optional<T>::and_then(F) requires the return type of F to be a specialization of optional "
+            "(N4901 [optional.monadic]/5).");
+
+        if (this->_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), _STD move(this->_Value));
+        } else {
+            return remove_cvref_t<_Uty>{};
+        }
+    }
+
+    template <class _Fn>
+    constexpr auto transform(_Fn&& _Func) & {
+        using _Uty = remove_cv_t<invoke_result_t<_Fn, _Ty&>>;
+
+        static_assert(!_Is_any_of_v<_Uty, nullopt_t, in_place_t>,
+            "optional<T>::transform(F) requires the return type of F to be a type other than nullopt_t or in_place_t "

```

Comment:
Technically "other than *cv*-`nullopt_t` or *cv*-`in_place_t`", but that's strictly worse. (Please don't change.)